### DISCUSSION TITLE
OCIO context variables, and drag-and-drop handling

### DISCUSSION THREAD

**Discussion by dbr** [ggarra13/mrv2#246] [Q&A]
Hello! I'm currently evaluating if we can replace our usage of RV with mrv2, and there's two important bits of functionality I couldn't tell if we can replicate, based on a quick look around the documentation:

First - we use OCIO for LUT's, and often have different grades or LUT's per shot, which we implement using context variables.

Currently in RV:
1. we have a Python plugin which listens to the `source-setup` event. It uses an internal Python API to parse the image sequence path, which gives the job/sequence/shot values (e.g `{'job': 'egg', 'seq': 'abc123', 'shot: 'abc123-321'}`
2. RV's OCIO nodes have a dictionary with the context variables names and values - our plugin sets the job/seq/shot values
3. Internally, RV passes these values into the a [Context](https://opencolorio.readthedocs.io/en/latest/api/context.html) and passes them to the current clip's `OCIO::Processor` instance

Is there a way to do the same dynamic (per-clip) OCIO context variables in mrv2?

Second, we use Flow (nee ShotGrid), and rely on the drag-and-drop handling to open clips for review.

For example into a session we would drag a link like `https://examplestudio.shotgunstudio.com/detail/Version/1234` - we would then use the `shotgun_api3` Python API to query the media info for that version ID, and create a new clip with that file (either a movie or frames). We also have a shortcut to toggle between the movie and frames. Same case with playlists URL's.

I couldn't see any examples of interacting with file-open events in the Python API - but it seems like something that might already be doable, and I'm just looking in the wrong place!

**Comment by ggarra13** [comment]
> Hello! I'm currently evaluating if we can replace our usage of RV with mrv2, and there's two important bits of functionality I couldn't tell if we can replicate, based on a quick look around the documentation:

Okay, first, try to download the latest beta of v1.1.9 from:

https://sourceforge.net/projects/mrv2/files/beta/

as there were some bugs in the LUTs and no plug-in mechanism to do what you want in v1.1.8.

> 
> First - we use OCIO for LUT's, and often have different grades or LUT's per shot, which we implement using context variables.
> 
> Currently in RV:
> 
> 1. we have a Python plugin which listens to the `source-setup` event. It uses an internal Python API to parse the image sequence path, which gives the job/sequence/shot values (e.g `{'job': 'egg', 'seq': 'abc123', 'shot: 'abc123-321'}`

v1.1.9's plug-ins added an on_open_file(fileName, audioFilename) method on Plug-ins which can be used for doing that.  You'd be responsible for choosing what ICS / LUTs on the Python code.  LUTs as well as OCIO ICS are stored with each clip when you switch from one clip to the next.  You can save a session which will store the LUT and OCIO ICS for all clips.

> 
> Is there a way to do the same dynamic (per-clip) OCIO context variables in mrv2?
> 

See above.

> Second, we use Flow (nee ShotGrid), and rely on the drag-and-drop handling to open clips for review.
> 

mrv2 supports playing https clips directly, like this one (it works on v1.1.8 command-line, no DnD):

https://dpel-assets.aswf.io/asc-stem2/ASC_StEM2_178_2K_24_100nits_Rec709_Stereo.mp4

Drag and drop functionality on v1.1.9 works fine on Windows, macOS and Linux's Firefox.  Currently there is a bug that makes it not work for Google Chrome.

> For example into a session we would drag a link like `https://examplestudio.shotgunstudio.com/detail/Version/1234` - we would then use the `shotgun_api3` Python API to query the media info for that version ID, and create a new clip with that file (either a movie or frames). We also have a shortcut to toggle between the movie and frames. Same case with playlists URL's.
> 

That, in principle, should be doable now in v1.1.9, and you might be able to play it directly without having to go throu the shotgun_api.
Unless, of course, Autodesk Flow encodes or encrypts the movie in some way, which makes it require the shotgun_api.

> I couldn't see any examples of interacting with file-open events in the Python API - but it seems like something that might already be doable, and I'm just looking in the wrong place!

No, you were not looking at the wrong place.  v1.1.8 did not offer the functionality.  It will come on v1.1.9, but you can test the beta and provide feedback.

Hope that helps!  I never used Shotgun, so if you have trouble getting it work, open an Issue and we'll put both of our minds together to make it work.
